models/User.js

```diff
@@ -0,0 +1,49 @@
+const mongoose = require('mongoose');
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+
+const UserSchema = new mongoose.Schema({
+  name: {
+    type: String,
+    required: [true, 'Please provide name'],
+    minlength: 3,
+    maxlength: 50,
+  },
+  email: {
+    type: String,
+    required: [true, 'Please provide email'],
+    match: [
+      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
+      'Please provide valid email',
+    ],
+    unique: true,
+  },
+  password: {
+    type: String,
+    required: [true, 'Please provide password'],
+    minlength: 6,
+  },
+});
+
+UserSchema.pre('save', async function (next) {
+  this.password = await bcrypt.hash(this.password, 10);
+
+  next();
+});
```

Comment:
Looks good.
Just wanted to share that Mongoose (version 5 and higher) gives us a nice shortcut: https://mongoosejs.com/docs/5.x/docs/middleware.html#pre

> In [mongoose 5.x](http://thecodebarbarian.com/introducing-mongoose-5.html#promises-and-async-await-with-middleware), instead of calling next() manually, you can use a function that returns a promise. In particular, you can use [async/await](http://thecodebarbarian.com/common-async-await-design-patterns-in-node.js.html).

So you could just omit the `next()` in this case since the only other statement in this function uses `await`

```suggestion
UserSchema.pre('save', async function (next) {
  this.password = await bcrypt.hash(this.password, 10);
});
```